frontend/src/types/dataSource.ts

```diff
@@ -15,15 +15,20 @@ import { Principal } from "./principal";
 // which from the ops perspective, having different meaning from the normal RW data source.
 export type DataSourceType = "ADMIN" | "RW" | "RO";
 
+export enum DataSourceTypes {
+  ADMIN = "ADMIN",
+  RW = "RW",
+  RO = "RO",
+}
+
 export type DataSource = {
   id: DataSourceId;
 
   // Related fields
+  databaseId: DatabaseId;
```

Comment:
Normally, we do not need to store databaseId / instanceId since client code accesses the full database/instance object

Reply:
Actually we don't have the full database/instance from backend api, but the databaseId/intanceId.
https://github.com/bytebase/bytebase/blob/main/api/data_source.go#L70

Reply:
OK, so then the convert method in dataSource.ts probably won't get the full database/instance at all since those must be included in the api response.

Then here we probably don't need to declare database/instance